File: packages/server/src/routes/messages.ts
import type { FastifyInstance } from 'fastify';
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { getDb } from '../db.js';
import * as Q from '../queries.js';
import { requirePermission } from '../middleware/permissions.js';
import { writeWorkspaceFileData } from './workspace.js';

export function registerMessageRoutes(app: FastifyInstance): void {
  // Get messages for a channel
  app.get<{
    Params: { channelId: string };
    Querystring: { before?: string; limit?: string; after?: string };
  }>('/api/v1/channels/:channelId/messages', async (request, reply) => {
    const { channelId } = request.params;
    const before = request.query.before ? parseInt(request.query.before, 10) : undefined;
    const after = request.query.after ? parseInt(request.query.after, 10) : undefined;
    const limit = request.query.limit ? Math.min(parseInt(request.query.limit, 10), 200) : 100;

    const db = getDb();
    const channel = Q.getChannel(db, channelId);
    if (!channel) {
      return reply.status(404).send({ error: 'Channel not found' });
    }

    if (channel.visibility === 'private') {
      const userId = request.currentUser?.id;
      if (!userId || !Q.canAccessChannel(db, channelId, userId)) {
        return reply.status(404).send({ error: 'Channel not found' });
      }
    }

    const result = Q.getMessages(db, channelId, before, limit, after);
    const messageIds = result.messages.map(m => m.id);
    const reactionsMap = Q.getReactionsForMessages(db, messageIds);
    const messagesWithReactions = result.messages.map(m => ({
      ...m,
      reactions: reactionsMap.get(m.id) ?? [],
    }));
    return { messages: messagesWithReactions, has_more: result.has_more };
  });

  // Search messages in a channel
  app.get<{
    Params: { channelId: string };
    Querystring: { q?: string };
  }>('/api/v1/channels/:channelId/messages/search', async (request, reply) => {
    const { channelId } = request.params;
    const q = request.query.q;

    if (!q || typeof q !== 'string' || q.trim().length === 0) {
      return reply.status(400).send({ error: 'Search query (q) is required' });
    }

    const db = getDb();
    const channel = Q.getChannel(db, channelId);
    if (!channel) {
      return reply.status(404).send({ error: 'Channel not found' });
    }

    if (channel.visibility === 'private') {
      const userId = request.currentUser?.id;
      if (!userId || !Q.canAccessChannel(db, channelId, userId)) {
        return reply.status(404).send({ error: 'Channel not found' });
      }
    }

    const messages = Q.searchMessages(db, channelId, q.trim(), 50);
    return { messages };
  });

  // Create message
  app.post<{
    Params: { channelId: string };
    Body: {
      content: string;
      content_type?: string;
      reply_to_id?: string;
      mentions?: string[];
    };
  }>('/api/v1/channels/:channelId/messages', { preHandler: [requirePermission('message.send', (req) => `channel:${(req.params as { channelId: string }).channelId}`)] }, async (request, reply) => {
    const { channelId } = request.params;
    const { content, content_type, reply_to_id, mentions } = request.body ?? {};

    if (!content || typeof content !== 'string' || content.trim().length === 0) {
      return reply.status(400).send({ error: 'Message content is required' });
    }

    const ct = content_type ?? 'text';
    if (ct !== 'text' && ct !== 'image') {
      return reply.status(400).send({ error: "content_type must be 'text' or 'image'" });
    }

    const db = getDb();
    const channel = Q.getChannel(db, channelId);
    if (!channel) {
      return reply.status(404).send({ error: 'Channel not found' });
    }

    const senderId = request.currentUser?.id;
    if (!senderId) {
      return reply.status(401).send({ error: 'Authentication required' });
    }

    if (channel.visibility === 'private' && !Q.canAccessChannel(db, channelId, senderId)) {
      return reply.status(404).send({ error: 'Channel not found' });
    }

    if (!Q.isChannelMember(db, channelId, senderId)) {
      return reply.status(403).send({ error: 'Not a member of this channel' });
    }

    const message = Q.createMessage(
      db,
      channelId,
      senderId,
      content.trim(),
      ct,
      reply_to_id ?? null,
      mentions ?? [],
    );

    const { broadcastToChannel } = await import('../ws.js');
    broadcastToChannel(channelId, {
      type: 'new_message',
      message,
    });

    if (ct === 'image' && content.startsWith('/uploads/')) {
      try {
        const uploadDir = process.env.UPLOAD_DIR ?? path.join(process.cwd(), 'data', 'uploads');
        const uploadFilename = content.replace('/uploads/', '');
        const uploadPath = path.join(uploadDir, uploadFilename);
        if (fs.existsSync(uploadPath)) {
          const attachmentsFolder = Q.mkdirWorkspace(db, senderId, channelId, null, 'attachments');
          const buffer = fs.readFileSync(uploadPath);
          const ext = path.extname(uploadFilename);
          const mimeMap: Record<string, string> = { '.jpg': 'image/jpeg', '.jpeg': 'image/jpeg', '.png': 'image/png', '.gif': 'image/gif', '.webp': 'image/webp' };
          const mime = mimeMap[ext] ?? 'application/octet-stream';
          const fileId = crypto.randomBytes(16).toString('hex');
          const fileName = `${message.id.slice(0, 8)}${ext}`;

          const siblings = Q.getSiblingNames(db, senderId, channelId, attachmentsFolder.id);
          const resolvedName = Q.resolveConflict(fileName, siblings);

          await writeWorkspaceFileData(senderId, channelId, fileId, buffer);
          Q.insertWorkspaceFile(db, {
            id: fileId,
            userId: senderId,
            channelId,
            parentId: attachmentsFolder.id,
            name: resolvedName,
            isDirectory: false,
            mimeType: mime,
            sizeBytes: buffer.length,
            source: 'message_attachment',
            sourceMessageId: message.id,
          });
        }
      } catch {
        // Non-critical: don't fail message send
      }
    }

    return reply.status(201).send({ message });
  });

  // Edit message
  app.put<{
    Params: { messageId: string };
    Body: { content: string };
  }>('/api/v1/messages/:messageId', async (request, reply) => {
    const { messageId } = request.params;
    const { content } = request.body ?? {};

    if (!content || typeof content !== 'string' || content.trim().length === 0) {
      return reply.status(400).send({ error: 'Content is required' });
    }

    const senderId = request.currentUser?.id;
    if (!senderId) {
      return reply.status(401).send({ error: 'Authentication required' });
    }

    const db = getDb();
    const existing = Q.getMessageById(db, messageId);
    if (!existing) {
      return reply.status(404).send({ error: 'Message not found' });
    }

    if (existing.deleted_at) {
      return reply.status(400).send({ error: 'Cannot edit deleted message' });
    }

    if (existing.sender_id !== senderId) {
      return reply.status(403).send({ error: 'Can only edit your own messages' });
    }

    const message = Q.updateMessageContent(db, messageId, content.trim());

    const { broadcastToChannel } = await import('../ws.js');
    broadcastToChannel(existing.channel_id, {
      type: 'message_edited',
      message,
    });

    const senderName = request.currentUser?.display_name ?? senderId;
    Q.insertEvent(db, 'message_edited', existing.channel_id, {
      ...message,
      sender_id: senderId,
      system_message: `用户 ${senderName} 编辑了消息`,
    });

    return { message };
  });

  // Delete message (soft delete)
  app.delete<{
    Params: { messageId: string };
  }>('/api/v1/messages/:messageId', async (request, reply) => {
    const { messageId } = request.params;

    const senderId = request.currentUser?.id;
    if (!senderId) {
      return reply.status(401).send({ error: 'Authentication required' });
    }

    const db = getDb();
    const existing = Q.getMessageById(db, messageId);
    if (!existing) {
      return reply.status(404).send({ error: 'Message not found' });
    }

    // Already deleted — idempotent
    if (existing.deleted_at) {
      return reply.status(204).send();
    }

    const isAdmin = request.currentUser?.role === 'admin';
    if (existing.sender_id !== senderId && !isAdmin) {
      return reply.status(403).send({ error: 'Permission denied' });
    }

    const { deleted_at } = Q.softDeleteMessage(db, messageId);

    const { broadcastToChannel } = await import('../ws.js');
    broadcastToChannel(existing.channel_id, {
      type: 'message_deleted',
      message_id: existing.id,
      channel_id: existing.channel_id,
      deleted_at,
    });

    const senderName = request.currentUser?.display_name ?? senderId;
    Q.insertEvent(db, 'message_deleted', existing.channel_id, {
      message_id: existing.id,
      channel_id: existing.channel_id,
      deleted_at,
      sender_id: senderId,
      system_message: `用户 ${senderName} 删除了一条消息`,
    });

    return reply.status(204).send();
  });
}

File: packages/server-go/internal/api/messages.go
package api

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"borgee-server/internal/auth"
	"borgee-server/internal/store"
)

// MessageHandler handles message CRUD endpoints.
//
// MentionDispatcher (DM-2.2, #312) is optional — nil at boot means
// mention parser + offline fallback are off (legacy / test paths).
// When non-nil, handleCreateMessage parses `@<uuid>` tokens, rejects
// cross-channel targets with 400 mention.target_not_in_channel, persists
// message_mentions rows (#361), and fans out mention_pushed frames or
// owner system DM fallback (acceptance §1.1-§2.5).
type MessageHandler struct {
	Store      *store.Store
	Logger     *slog.Logger
	Hub        EventBroadcaster
	Mentions   *MentionDispatcher
}

type EventBroadcaster interface {
	BroadcastEventToChannel(channelID string, eventType string, payload any)
	BroadcastEventToAll(eventType string, payload any)
	BroadcastEventToUser(userID string, eventType string, payload any)
	SignalNewEvents()
}

func (h *MessageHandler) RegisterRoutes(mux *http.ServeMux, authMw func(http.Handler) http.Handler, sendPerm func(http.Handler) http.Handler, readPerm func(http.Handler) http.Handler) {
	// Channel-scoped routes (need auth)
	// AP-0-bis: GET /channels/:id/messages now requires `message.read`
	// capability. New agents get this in default grants; legacy agents get
	// backfilled by migration v=8 (ap_0_bis_message_read.go). Reverse
	// assertion: agent without message.read row → 403 (see messages_perm_test.go).
	mux.Handle("GET /api/v1/channels/{channelId}/messages", authMw(readPerm(http.HandlerFunc(h.handleListMessages))))
	mux.Handle("GET /api/v1/channels/{channelId}/messages/search", authMw(readPerm(http.HandlerFunc(h.handleSearchMessages))))
	mux.Handle("POST /api/v1/channels/{channelId}/messages", authMw(sendPerm(http.HandlerFunc(h.handleCreateMessage))))

	// Message-scoped routes (need auth)
	mux.Handle("PUT /api/v1/messages/{messageId}", authMw(http.HandlerFunc(h.handleUpdateMessage)))
	mux.Handle("DELETE /api/v1/messages/{messageId}", authMw(http.HandlerFunc(h.handleDeleteMessage)))
}

// GET /api/v1/channels/:channelId/messages?before=&after=&limit=
func (h *MessageHandler) handleListMessages(w http.ResponseWriter, r *http.Request) {
	channelID := r.PathValue("channelId")
	if channelID == "" {
		writeJSONError(w, http.StatusBadRequest, "Channel ID is required")
		return
	}

	user := auth.UserFromContext(r.Context())

	// Validate channel exists
	ch, err := h.Store.GetChannelByID(channelID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	// Private channel access check
	if ch.Visibility == "private" {
		if user == nil || !h.Store.CanAccessChannel(channelID, user.ID) {
			writeJSONError(w, http.StatusNotFound, "Channel not found")
			return
		}
	}

	// Parse query params
	var before, after *int64
	if v := r.URL.Query().Get("before"); v != "" {
		if n, err := strconv.ParseInt(v, 10, 64); err == nil {
			before = &n
		}
	}
	if v := r.URL.Query().Get("after"); v != "" {
		if n, err := strconv.ParseInt(v, 10, 64); err == nil {
			after = &n
		}
	}

	limit := 100
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			if n > 200 {
				n = 200
			}
			limit = n
		}
	}

	msgs, hasMore, err := h.Store.ListChannelMessages(channelID, before, after, limit)
	if err != nil {
		h.Logger.Error("failed to list messages", "error", err)
		writeJSONError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	type messageWithReactions struct {
		store.MessageWithSender
		Reactions []store.AggregatedReaction `json:"reactions"`
	}
	// TODO: N+1 query — each message triggers a separate DB query for reactions.
	// Optimize with batch query: SELECT ... WHERE message_id IN (...) grouped by message_id.
	out := make([]messageWithReactions, len(msgs))
	for i, msg := range msgs {
		reactions, err := h.Store.GetReactionsByMessage(msg.ID)
		if err != nil {
			h.Logger.Error("failed to get message reactions", "error", err)
			writeJSONError(w, http.StatusInternalServerError, "Internal server error")
			return
		}
		if reactions == nil {
			reactions = []store.AggregatedReaction{}
		}
		out[i] = messageWithReactions{MessageWithSender: msg, Reactions: reactions}
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{
		"messages": out,
		"has_more": hasMore,
	})
}

// GET /api/v1/channels/:channelId/messages/search?q=&limit=
func (h *MessageHandler) handleSearchMessages(w http.ResponseWriter, r *http.Request) {
	channelID := r.PathValue("channelId")
	if channelID == "" {
		writeJSONError(w, http.StatusBadRequest, "Channel ID is required")
		return
	}

	user := auth.UserFromContext(r.Context())

	// Validate channel exists
	ch, err := h.Store.GetChannelByID(channelID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	if ch.Visibility == "private" {
		if user == nil || !h.Store.CanAccessChannel(channelID, user.ID) {
			writeJSONError(w, http.StatusNotFound, "Channel not found")
			return
		}
	}

	q := strings.TrimSpace(r.URL.Query().Get("q"))
	if q == "" {
		writeJSONError(w, http.StatusBadRequest, "Search query (q) is required")
		return
	}

	limit := 50
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			if n > 50 {
				n = 50
			}
			limit = n
		}
	}

	msgs, err := h.Store.SearchMessages(channelID, q, limit)
	if err != nil {
		h.Logger.Error("failed to search messages", "error", err)
		writeJSONError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"messages": msgs})
}

// POST /api/v1/channels/:channelId/messages
func (h *MessageHandler) handleCreateMessage(w http.ResponseWriter, r *http.Request) {
	channelID := r.PathValue("channelId")
	if channelID == "" {
		writeJSONError(w, http.StatusBadRequest, "Channel ID is required")
		return
	}

	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	var body struct {
		Content     string   `json:"content"`
		ContentType string   `json:"content_type"`
		ReplyToID   *string  `json:"reply_to_id"`
		Mentions    []string `json:"mentions"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSONError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	content := strings.TrimSpace(body.Content)
	if content == "" {
		writeJSONError(w, http.StatusBadRequest, "Message content is required")
		return
	}

	ct := body.ContentType
	if ct == "" {
		ct = "text"
	}
	if ct != "text" && ct != "image" && ct != "command" && ct != "artifact_comment" {
		writeJSONError(w, http.StatusBadRequest, "content_type must be 'text', 'image', 'command', or 'artifact_comment'")
		return
	}

	// Validate channel exists
	ch, err := h.Store.GetChannelByID(channelID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	// Private channel access check
	if ch.Visibility == "private" && !h.Store.CanAccessChannel(channelID, user.ID) {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	// Must be a member
	if !h.Store.IsChannelMember(channelID, user.ID) {
		writeJSONError(w, http.StatusForbidden, "Not a member of this channel")
		return
	}

	// DM-2.2 mention parser + cross-channel guard.
	// 立场 ① parse `@<uuid>` token 拆死 raw UUID; 立场 ② cross-channel
	// reject 400 (mention 仅在 channel 内, 跟 RT-1/CHN-1 留账边界对齐).
	// Dispatch (online push / offline owner DM fallback) 在 message
	// 落库后执行 — 失败仅 log 不阻断 message 创建 (best-effort fanout,
	// 反约束 (§2.4): fallback 走 owner 后台不污染发送方).
	var parsedMentionTargets []string
	if h.Mentions != nil {
		targets, offender, mErr := h.Mentions.MentionTargetsFromBody(channelID, content)
		if mErr != nil {
			// ErrMentionTargetNotInChannel → 400 with offender hint.
			writeJSONError(w, http.StatusBadRequest, "mention.target_not_in_channel: "+offender)
			return
		}
		parsedMentionTargets = targets
	}

	// CV-8: artifact comment thread reply validators (1-level depth + agent
	// thinking subject 5-pattern 第 6 处链 RT-3 + BPP-2.2 + AL-1b + CV-5 +
	// CV-7 + CV-8 byte-identical). Only fires on artifact_comment with
	// reply_to_id; non-comment paths unchanged.
	if ct == "artifact_comment" && body.ReplyToID != nil && *body.ReplyToID != "" {
		parent, perr := h.Store.GetMessageByID(*body.ReplyToID)
		if perr != nil {
			writeJSONErrorCode(w, http.StatusBadRequest, "comment.reply_target_invalid", "reply target not found")
			return
		}
		if parent.ContentType != "artifact_comment" {
			writeJSONErrorCode(w, http.StatusBadRequest, "comment.reply_target_invalid", "reply target must be an artifact comment")
			return
		}
		if parent.ReplyToID != nil && *parent.ReplyToID != "" {
			writeJSONErrorCode(w, http.StatusBadRequest, "comment.thread_depth_exceeded", "thread depth limited to 1 level")
			return
		}
		// Agent senders must pass the 5-pattern thinking-subject guard
		// byte-identical to CV-5 / CV-7 (errcode 同字符串).
		if user.Role == "agent" && violatesThinkingSubjectCV8(content) {
			writeJSONErrorCode(w, http.StatusBadRequest, "comment.thinking_subject_required",
				"agent comment must carry a concrete subject (thinking-only body rejected)")
			return
		}
	}

	msg, err := h.Store.CreateMessageFull(channelID, user.ID, content, ct, body.ReplyToID, body.Mentions)
	if err != nil {
		h.Logger.Error("failed to create message", "error", err)
		writeJSONError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	// DM-2.2 persist + dispatch — non-blocking on errors (log + continue).
	// PersistMentions writes #361 message_mentions rows; Dispatch fans
	// online targets via PushMentionPushed and offline agents via owner
	// system DM (#314 §1 ③ byte-identical body).
	if h.Mentions != nil && len(parsedMentionTargets) > 0 {
		if pErr := h.Mentions.PersistMentions(msg.ID, parsedMentionTargets); pErr != nil {
			h.Logger.Error("failed to persist mentions", "error", pErr, "message_id", msg.ID)
		}
		if dErr := h.Mentions.Dispatch(msg.ID, channelID, ch.Name, user.ID, content, parsedMentionTargets, msg.CreatedAt); dErr != nil {
			h.Logger.Error("mention dispatch partial failure", "error", dErr, "message_id", msg.ID)
		}
	}

	h.Store.CreateEvent(&store.Event{
		Kind:      "new_message",
		ChannelID: channelID,
		Payload:   mustJSON(map[string]any{"message": msg}),
	})
	if h.Hub != nil {
		h.Hub.BroadcastEventToChannel(channelID, "new_message", map[string]any{"message": msg})
	}

	writeJSONResponse(w, http.StatusCreated, map[string]any{"message": msg})
}

// PUT /api/v1/messages/:messageId
func (h *MessageHandler) handleUpdateMessage(w http.ResponseWriter, r *http.Request) {
	messageID := r.PathValue("messageId")
	if messageID == "" {
		writeJSONError(w, http.StatusBadRequest, "Message ID is required")
		return
	}

	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	var body struct {
		Content string `json:"content"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSONError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	content := strings.TrimSpace(body.Content)
	if content == "" {
		writeJSONError(w, http.StatusBadRequest, "Content is required")
		return
	}

	existing, err := h.Store.GetMessageByID(messageID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Message not found")
		return
	}

	if existing.DeletedAt != nil {
		writeJSONError(w, http.StatusBadRequest, "Cannot edit deleted message")
		return
	}

	// CM-3.2: cross-org 403 (#200 §3 row 1).
	// MUST run BEFORE the AP-5 channel-member gate — cross-org contract
	// (cm-3-resource-ownership-checklist.md §3) explicitly returns 403 for
	// foreign-org callers (TestCrossOrgRead403 lock).
	if store.CrossOrg(user.OrgID, existing.OrgID) {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	// AP-5 立场 ① — channel-member ACL gate (跟 AP-4 reactions 同模式).
	// Closes the post-removal gap: a same-org user removed from the
	// channel must not be able to edit messages they previously sent
	// there. byte-identical "Channel not found" 404 fail-closed.
	if !h.Store.IsChannelMember(existing.ChannelID, user.ID) || !h.Store.CanAccessChannel(existing.ChannelID, user.ID) {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	if existing.SenderID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Can only edit your own messages")
		return
	}

	// CV-7 立场 ③: agent edit on artifact-comment-typed message must
	// re-pass the 5-pattern thinking-subject guard. byte-identical 跟
	// CV-5 #530 artifact_comments.go::violatesThinkingSubject — 5-pattern
	// 第 5 处链 (RT-3 + BPP-2.2 + AL-1b + CV-5 + CV-7). 5-pattern 改 =
	// 改 5 处 byte-identical.
	if existing.ContentType == "artifact_comment" {
		sender, _ := h.Store.GetUserByID(existing.SenderID)
		if sender != nil && sender.Role == "agent" && violatesThinkingSubjectCV7(content) {
			writeJSONErrorCode(w, http.StatusBadRequest, "comment.thinking_subject_required",
				"agent comment must carry a concrete subject (thinking-only body rejected)")
			return
		}
	}

	msg, err := h.Store.UpdateMessage(messageID, content)
	if err != nil {
		h.Logger.Error("failed to update message", "error", err)
		writeJSONError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	// Write edit event
	h.Store.CreateEvent(&store.Event{
		Kind:      "message_edited",
		ChannelID: existing.ChannelID,
		Payload:   mustJSON(map[string]any{"id": messageID, "channel_id": existing.ChannelID, "sender_id": user.ID, "content": content, "system_message": "用户 " + user.DisplayName + " 编辑了消息"}),
	})
	if h.Hub != nil {
		h.Hub.BroadcastEventToChannel(existing.ChannelID, "message_edited", map[string]any{"message": msg})
	}

	writeJSONResponse(w, http.StatusOK, map[string]any{"message": msg})
}

// DELETE /api/v1/messages/:messageId
func (h *MessageHandler) handleDeleteMessage(w http.ResponseWriter, r *http.Request) {
	messageID := r.PathValue("messageId")
	if messageID == "" {
		writeJSONError(w, http.StatusBadRequest, "Message ID is required")
		return
	}

	user := auth.UserFromContext(r.Context())
	if user == nil {
		writeJSONError(w, http.StatusUnauthorized, "Unauthorized")
		return
	}

	existing, err := h.Store.GetMessageByID(messageID)
	if err != nil {
		writeJSONError(w, http.StatusNotFound, "Message not found")
		return
	}

	// Already deleted — idempotent
	if existing.DeletedAt != nil {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	// CM-3.2: cross-org 403. MUST run BEFORE the AP-5 channel-member gate
	// — cross-org contract returns 403 for foreign-org callers
	// (TestCrossOrgRead403 lock).
	if store.CrossOrg(user.OrgID, existing.OrgID) {
		writeJSONError(w, http.StatusForbidden, "Forbidden")
		return
	}

	// AP-5 立场 ① — channel-member ACL gate (跟 AP-4 + handleUpdateMessage 同模式).
	// Closes the post-removal gap on DELETE.
	if !h.Store.IsChannelMember(existing.ChannelID, user.ID) || !h.Store.CanAccessChannel(existing.ChannelID, user.ID) {
		writeJSONError(w, http.StatusNotFound, "Channel not found")
		return
	}

	// ADM-0.3: no role short-circuit. Sender-only delete on the user-rail;
	// admin-rail message delete uses /admin-api/v1/messages.
	if existing.SenderID != user.ID {
		writeJSONError(w, http.StatusForbidden, "Permission denied")
		return
	}

	deletedAt, err := h.Store.SoftDeleteMessage(messageID)
	if err != nil {
		h.Logger.Error("failed to delete message", "error", err)
		writeJSONError(w, http.StatusInternalServerError, "Internal server error")
		return
	}

	// Write delete event
	h.Store.CreateEvent(&store.Event{
		Kind:      "message_deleted",
		ChannelID: existing.ChannelID,
		Payload:   mustJSON(map[string]any{"message_id": messageID, "channel_id": existing.ChannelID, "deleted_at": deletedAt, "sender_id": user.ID, "system_message": "用户 " + user.DisplayName + " 删除了一条消息"}),
	})
	if h.Hub != nil {
		h.Hub.BroadcastEventToChannel(existing.ChannelID, "message_deleted", map[string]any{"message_id": messageID, "channel_id": existing.ChannelID, "deleted_at": deletedAt})
	}

	w.WriteHeader(http.StatusNoContent)
}

// mustJSON marshals v to JSON string, returning "{}" on error.
func mustJSON(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		return "{}"
	}
	return string(b)
}
